plugins/experimental/alt_svc/single_service_file_map.cc

```diff
@@ -0,0 +1,115 @@
+/**
+  @file
+  @brief A particular IPHostMap implementation which takes a static file and routes client IPs based on that file.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include "ip_host_map.h"
+#include "prefix_parser.h"
+#include <fstream>
+#include <algorithm>
+
+using namespace std;
+
+const string ssfm_empty = "";
+
+char *
+SingleServiceFileMap::findHostForIP(const sockaddr *ip) const noexcept
+{
+  char *data = nullptr;
+  this->host_map.contains(ip, (void **)&data);
+  return data;
+}
+
+bool
+SingleServiceFileMap::isValid() const noexcept
+{
+  return _isValid;
+}
+
+// Lifted from ControlMatcher.cc
+void
+SingleServiceFileMap::print_the_map() const noexcept
+{
+  TS_DEBUG(PLUGIN_NAME, "\tIp Matcher with %zu ranges.\n", this->host_map.getCount());
+  for (auto &spot : this->host_map) {
+    char b1[INET6_ADDRSTRLEN], b2[INET6_ADDRSTRLEN];
+    TS_DEBUG(PLUGIN_NAME, "\tRange %s - %s ", ats_ip_ntop(spot.min(), b1, sizeof b1), ats_ip_ntop(spot.max(), b2, sizeof b2));
+    TS_DEBUG(PLUGIN_NAME, "Host: %s \n", static_cast<char *>(spot.data()));
+  }
+}
+
+SingleServiceFileMap::SingleServiceFileMap(ts::string_view filename)
+{
+  // Read file
+  bool fail = 0;
+  ifstream config_file{filename.data()};
+  if (config_file.fail()) {
+    TS_DEBUG(PLUGIN_NAME, "Cannot find a config file at: %s", filename.data());
+    fail = 1;
+  } else {
+    // Parse file into plugin-local IpMap
+    string ip_with_prefix, buff;
+    char *hostname = nullptr; // TODO can we make this a string?
+    while (!getline(config_file, buff).eof()) {
+      bool is_host = (buff[0] != ' ');
+      buff.erase(remove_if(buff.begin(), buff.end(), ::isspace), buff.end());
+      if (is_host) {
+        hostname = static_cast<char *>(malloc(buff.size() + 1));
```

Comment:
I don't know how I can avoid this memory leak. The hostname is never de-allocated, and basically exists as long as the IpMap. I want all the IP ranges to point to the same ever-existing hostname. This is only done on start-up, but is there a better way to do that?